src/main/java/org/tron/common/runtime/RuntimeImpl.java

```diff
@@ -270,25 +266,81 @@ private long getEnergyLimit(AccountCapsule creator, AccountCapsule caller,
     }
   }
 
-  private double getThisTxCPULimitInUsRatio() {
+  public long getEnergyLimitWithFixRatio(AccountCapsule creator, AccountCapsule caller,
+      TriggerSmartContract contract, long feeLimit, long callValue) {
+
+    long callerEnergyLimit = getEnergyLimitWithFixRatio(caller, feeLimit, callValue);
+    if (Arrays.equals(creator.getAddress().toByteArray(), caller.getAddress().toByteArray())) {
+      // when the creator calls his own contract, this logic will be used.
+      // so, the creator must use a BIG feeLimit to call his own contract,
+      // which will cost the feeLimit TRX when the creator's frozen energy is 0.
+      return callerEnergyLimit;
+    }
+
+    long creatorEnergyLimit = 0;
+    ContractCapsule contractCapsule = this.deposit
+        .getContract(contract.getContractAddress().toByteArray());
+    long consumeUserResourcePercent = contractCapsule.getConsumeUserResourcePercent();
+
+    long originEnergyLimit = contractCapsule.getOriginEnergyLimit();
+
+    if (consumeUserResourcePercent <= 0) {
+      creatorEnergyLimit = min(energyProcessor.getAccountLeftEnergyFromFreeze(creator),
+          originEnergyLimit);
+    } else if (consumeUserResourcePercent < Constant.ONE_HUNDRED) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: "if ... else if" constructs should end with "else" clauses. ([squid:S126](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS126))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1670&issues=AWbSdHUuSgN4uCclEWm-&open=AWbSdHUuSgN4uCclEWm-)

[ISSUE_KEY]: AWbSdHUuSgN4uCclEWm-